airbyte-integrations/connectors/source-google-analytics-v4/source_google_analytics_v4/source.py

```diff
@@ -144,10 +157,60 @@ def request_body_json(
 
         return request_body
 
+    def get_json_schema(self) -> Mapping[str, Any]:
+        """
+        Override get_json_schema CDK method to retrieve the schema information for GoogleAnalyticsV4 Object dynamically.
+        """
+
+        schema = {"type": ["null", "object"], "additionalProperties": False, "properties": {"view_id": {"type": ["string"]}}}
+
+        # Add the dimensions to the schema
+        for dimension in self.report["dimensions"]:
+            data_type = self.lookup_data_type("dimension", dimension)
+            dimension = dimension.replace("ga:", "ga_")
+
+            schema["properties"][dimension] = {
+                "type": [data_type],
+            }
+
+        # Add the metrics to the schema
+        for metric in self.report["metrics"]:
+            data_type = self.lookup_data_type("metric", metric)
+            metric = metric.replace("ga:", "ga_")
+
+            schema["properties"][metric] = {
+                # metrics are allowed to also have null values
+                "type": ["null", data_type],
+            }
+
+        return schema
+
+    def stream_slices(self, stream_state: Mapping[str, Any] = None, **kwargs) -> Iterable[Optional[Mapping[str, Any]]]:
```

Comment:
For example

start date     = 2020.01.01
window_in_days = 90 days (3 month)

it means that we will have about 7 slices, like this (not exactly like this):

2020.01.01 - 2020.04.01
2020.04.02 - 2020.07.01
2020.07.02 - 2020.10.01
2020.10.02 - 2020.12.01
2021.01.01 - 2021.04.01
2021.04.02 - 2021.07.01
2020.07.02 - NOW

Reply:
as far as I can see there is always only one slice

Reply:
yes, only one slice start date to end date - first sync, state to the new end date (state + window size) - second sync

Reply:
Please follow the example in my previous message.

window_in_days - is the size of the slice. So we do expect to have 7 slices in provided examples.

This is the last 'opened' comment I have for this review. Once it is fixed we can proceed with external review.

Reply:
Updated

Reply:
slices should not overlap, otherwise we could sync the same data a few times.

Reply:
now end_date is a start_date for next slice. But should like this if  end_date 2021-01-01 then next start_date should be 2021-01-02